Positive and Negative Trait Implementations Clash Unexpectedly

### Code

```Rust
#![allow(unused)]
#![feature(negative_impls)]

trait MyTrait {}

#[derive(Clone)]
struct MyString {
    string: String,
}

impl<T: Copy> !MyTrait for T { }

// Works
impl MyTrait for MyString { }
// Throws error but it should work
impl MyTrait for String { }

fn main() {
    
}
```

`rustc --version --verbose`:
```
rustc 1.85.0-nightly (5f23ef7d3 2024-12-20)
binary: rustc
commit-hash: 5f23ef7d3f7a8c3e0ca5c4e1978829c0448a3686
commit-date: 2024-12-20
host: aarch64-unknown-linux-gnu
release: 1.85.0-nightly
LLVM version: 19.1.6
```

### Error output

```
error[E0751]: found both positive and negative implementation of trait `MyTrait` for type `String`:
  --> src/main.rs:16:1
   |
11 | impl<T: Copy> !MyTrait for T { }
   | ---------------------------- negative implementation here
...
16 | impl MyTrait for String { }
   | ^^^^^^^^^^^^^^^^^^^^^^^ positive implementation here
```

So we aren't as good at explaining coherence errors for negative impls, but if you change this to a *positive* impl you'll get the justification you need:

```
error[E0119]: conflicting implementations of trait `MyTrait` for type `std::string::String`
 --> <source>:8:1
  |
5 | impl<T: Copy> MyTrait for T { }
  | --------------------------- first implementation here
...
8 | impl MyTrait for String { }
  | ^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `std::string::String`
  |
  = note: upstream crates may add a new impl of trait `std::marker::Copy` for type `std::string::String` in future versions
```

That is to say, according to the orphan rules, the standard library is is *technically* free to add a new `impl Copy for String` implementation in the future. This would break your downstream crate, so the compiler proactively rejects this impl overlap.

The reason it works for `MyString` is because that's a type that is local to your crate.

--- a/compiler/rustc_trait_selection/src/traits/specialize/mod.rs
+++ b/compiler/rustc_trait_selection/src/traits/specialize/mod.rs
@@ -483,15 +483,19 @@ fn report_negative_positive_conflict<'tcx>(
     negative_impl_def_id: DefId,
     positive_impl_def_id: DefId,
 ) -> ErrorGuaranteed {
-    tcx.dcx()
-        .create_err(NegativePositiveConflict {
-            impl_span: tcx.def_span(local_impl_def_id),
-            trait_desc: overlap.trait_ref,
-            self_ty: overlap.self_ty,
-            negative_impl_span: tcx.span_of_impl(negative_impl_def_id),
-            positive_impl_span: tcx.span_of_impl(positive_impl_def_id),
-        })
-        .emit()
+    let mut diag = tcx.dcx().create_err(NegativePositiveConflict {
+        impl_span: tcx.def_span(local_impl_def_id),
+        trait_desc: overlap.trait_ref,
+        self_ty: overlap.self_ty,
+        negative_impl_span: tcx.span_of_impl(negative_impl_def_id),
+        positive_impl_span: tcx.span_of_impl(positive_impl_def_id),
+    });
+
+    for cause in &overlap.intercrate_ambiguity_causes {
+        cause.add_intercrate_ambiguity_hint(&mut diag);
+    }
+
+    diag.emit()
 }
 
 fn report_conflicting_impls<'tcx>(

--- a/tests/ui/traits/negative-impls/ambiguity-cause.negative_coherence.stderr
+++ b/tests/ui/traits/negative-impls/ambiguity-cause.negative_coherence.stderr
@@ -1,0 +1,14 @@
+error[E0119]: conflicting implementations of trait `MyTrait` for type `String`
+  --> $DIR/ambiguity-cause.rs:10:1
+   |
+LL | impl<T: Copy> MyTrait for T { }
+   | --------------------------- first implementation here
+LL |
+LL | impl MyTrait for String { }
+   | ^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `String`
+   |
+   = note: upstream crates may add a new impl of trait `std::marker::Copy` for type `std::string::String` in future versions
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0119`.

--- a/tests/ui/traits/negative-impls/ambiguity-cause.rs
+++ b/tests/ui/traits/negative-impls/ambiguity-cause.rs
@@ -1,0 +1,13 @@
+//@ revisions: simple negative_coherence
+
+#![feature(negative_impls)]
+#![cfg_attr(negative_coherence, feature(with_negative_coherence))]
+
+trait MyTrait {}
+
+impl<T: Copy> MyTrait for T { }
+
+impl MyTrait for String { }
+//~^ ERROR conflicting implementations of trait `MyTrait` for type `String`
+
+fn main() {}

--- a/tests/ui/traits/negative-impls/ambiguity-cause.simple.stderr
+++ b/tests/ui/traits/negative-impls/ambiguity-cause.simple.stderr
@@ -1,0 +1,14 @@
+error[E0119]: conflicting implementations of trait `MyTrait` for type `String`
+  --> $DIR/ambiguity-cause.rs:10:1
+   |
+LL | impl<T: Copy> MyTrait for T { }
+   | --------------------------- first implementation here
+LL |
+LL | impl MyTrait for String { }
+   | ^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `String`
+   |
+   = note: upstream crates may add a new impl of trait `std::marker::Copy` for type `std::string::String` in future versions
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0119`.

